src/TestFramework/PhpUnit/Coverage/IndexXmlCoverageParser.php

```diff
@@ -74,50 +59,35 @@ public function __construct(string $coverageDir)
      *
      * @throws NoLineExecuted
      *
-     * @return CoverageFileData[]
+     * @return iterable<CoveredFileData>
      */
-    public function parse(string $coverageXmlContent): array
+    public function parse(string $coverageXmlContent): iterable
     {
-        $xPath = self::createXPath($coverageXmlContent);
+        $xPath = XPathFactory::createXPath($coverageXmlContent);
 
         self::assertHasCoverage($xPath);
 
-        $nodes = $xPath->query('//file');
-
-        $projectSource = self::getProjectSource($xPath);
-
-        $data = [];
-
-        foreach ($nodes as $node) {
-            $relativeFilePath = $node->getAttribute('href');
-
-            $this->processCoverageFile($relativeFilePath, $projectSource, $data);
-        }
-
-        return $data;
+        return $this->parseNodes($xPath);
     }
 
-    private static function createXPath(string $coverageContent): SafeDOMXPath
+    /**
+     * @return iterable<CoveredFileData>
+     */
+    private function parseNodes(SafeDOMXPath $xPath): iterable
     {
-        $document = new DOMDocument();
-        $success = @$document->loadXML(self::removeNamespace($coverageContent));
+        $projectSource = self::getProjectSource($xPath);
 
-        Assert::true($success);
+        $nodes = $xPath->query('//file');
 
-        return new SafeDOMXPath($document);
-    }
+        foreach ($nodes as $node) {
+            $relativeCoverageFilePath = $node->getAttribute('href');
 
-    /**
-     * Remove namespace to work with xPath without a headache
-     */
-    private static function removeNamespace(string $xml): string
-    {
-        /** @var string $cleanedXml */
-        $cleanedXml = preg_replace('/xmlns=\".*?\"/', '', $xml);
+            $fileInfoProvider = new SourceFileInfoProvider($this->coverageDir, $relativeCoverageFilePath, $projectSource);
 
-        Assert::string($cleanedXml);
+            $parser = new XmlCoverageParser($fileInfoProvider);
```

Comment:
What do you think about creating this class 1 time before the loop and pass `$fileInfoProvider` to the `parse()` method?

Reply:
It is really hard to see it from the diff here, but $fileInfoProvider is also [called once per node](https://github.com/infection/infection/blob/bf44abc1ed5c94998ef7d7490c443d30f6069496/src/TestFramework/PhpUnit/Coverage/IndexXmlCoverageParser.php#L82-L89).